File: Practicas/src/main/java/ejerciciosexamen/monitor/drakkar.cpp
#include <condition_variable>
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

class drakkar {
  inline static int cant{0};
  std::mutex mutex;
  std::condition_variable cv;

 public:
  drakkar() = default;
  void cocinar() {
    std::unique_lock<std::mutex> lock(mutex);
    cant += 10;
    std::cout << cant << std::endl;
    cv.notify_all();
    while (cant > 0) cv.wait(lock);
  }

  void comer() {
    std::unique_lock<std::mutex> lock(mutex);
    if (cant > 0) {
      std::cout << cant << std::endl;
      --cant;
    } else {
      cv.notify_one();
      cv.wait(lock);
    }
  }
};

void run(drakkar& d, int tipo) {
  while (true) {
    if (tipo == 0)
      d.comer();
    else
      d.cocinar();
  }
}

int main() {
  drakkar a;
  std::vector<std::thread> threads;
  int cores = std::thread::hardware_concurrency();
  for (size_t i = 0; i < cores; i++) {
    threads.push_back(std::thread(run, std::ref(a), i));
  }
  for (auto& t : threads) {
    t.join();
  }
}

File: Practicas/src/main/java/ejerciciosexamen/monitor/drakkar.java
package ejerciciosexamen.monitor;

/*
La tripulación de un drakkar vikingo comparte una marmita con un al-
muerzo a base de m anguilas cocinadas al eneldo. Cuando un vikingo quiere
comer, se sirve una anguila. Si ya no quedan, avisa al vikingo cocinero pa-
ra que proceda a llenar la marmita de nuevo, utilizando las inagotables pro-
visiones de anguilas disponibles. Desarrolle un programa en java que modele
esta curiosa situación, y provea la sincronización necesaria utilizando méto-
dos synchronized y notificacion wait()-notifyAll(). Llame a su programa
drakkarVikingo.java.
*/
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class drakkar implements Runnable {
    private int tipo = 0;
    public static monitor m = new monitor();

    public drakkar(int tipo) {
        this.tipo = tipo;
    }

    public static void main(String args[]) {
        int cores = Runtime.getRuntime().availableProcessors();
        ExecutorService exe = Executors.newFixedThreadPool(cores);
        for (int i = 0; i < cores; i++) {
            exe.execute(new drakkar(i));
        }
        exe.shutdown();
        try {
            exe.awaitTermination(1, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while (true) {
            if (tipo == 0)
                m.rellenar();
            else
                m.comer();
        }
    }
}

class monitor {
    private static int cant = 0;

    public synchronized void comer() {
        try {
            if (cant > 0) {
                System.out.println("como");
                cant -= 1;
                System.out.println(cant);
            } else {
                notify();
                wait();
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public synchronized void rellenar() {
        cant += 10;
        System.out.println("rellena");
        System.out.println(cant);
        notifyAll();
        try {
            wait();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}